### DISCUSSION TITLE
What happens if I run an async func without await, in api route?

### DISCUSSION THREAD

**Discussion by oran1248** [vercel/next.js#19199] [Help]
Hi,

I'm wondering what will happen if I run something like this on one of my api routes:
```
const handler = async (req, res) => {
      someLogToDbAsyncFunc();
      return res.status(200).send();
}
```

So `someLogToDbAsyncFunc` is an async function that logs data to a mongo db.  
I don't want this operation to block the user from getting the response, so I omit the `await` keyword before the function call.
I want to make sure that in this case, the handler will terminated only after it finishes to run `someLogAsyncFunc` and indeed it will not block the response to the user.

Am I right here? It is ok to do that?

**Comment by marc-on-github** [comment]
Did you figure that out? I'm asking myself the same question right now. @oran1248

**Comment by henrikvtcodes** [comment]
So what you're looking for is
make db call --> send http response --> terminate handler function (only after the promise resolves)  
Assuming this is what you're looking for, you'd likely want to use `Promise.all()` or `Promise.allSettled()`. More information can be found [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#static_methods)

**Comment by icyJoseph** [comment]
I'd guess this is an instance of a memory leak that works on your favour. Deploys on Vercel do have a time limit on Serverless Functions, which starts at 10 seconds. 

What do I mean by memory leak?

Let's refactor the original code slightly:

```ts
const handler = async (req, res) => {
     // assume we can chain `.then` into this asynchronous work
      someLogToDbAsyncFunc().then(()=> {
        // this surely happens after we return
        sleep(1000).then(() => console.log("Deep sleep"));
      });

      return res.status(200).send();
}
```

The way asynchronous works, `someLogToDbAsyncFunc` (assuming it returns a promise) gets added to the event loop, means that the callback will eventually run, and then its nested callback once again. Once the synchronous code is done for, the event loop is processed, and the node process waits for it. This is the standard behaviour anyway.

Leaving open handlers, unterminated timers, unresolved promises, etc., are all forms of memory leaks. 

Like I said, the deploy environment might kill the lambda before it finishes its work, typically after 10 seconds. 

You can always account each call to your logger, by synchronously placing a marker (a log statement) indicating that the work will be started, perhaps even return a job id, and pass that id to the async work, which on termination should log that the job was finished.

I know some people use the `async_hooks` API from Node.js to do this accounting and keep track of separate asynchronous contexts, which allows you to natively to the above.

**Reply by gerzon-everneat** [reply]
the goal is not to add additional overhead, run the promise synchronously without putting await.

**Reply by brunobely** [reply]
Given that it's neatly cleaned up after the callback runs, I'd hesitate to call it a memory leak.

Logging something in a fire-and-forget manner (same as using `after`, but just by firing a promise with a callback) seems like a valid use case if you don't want the logging API to take up request time.

**Reply by icyJoseph** [reply]
Yeah, you are correct. I can't remember why I framed it like that. I think at the time I was debugging a somewhat similar issue, and ended up writing the answer, to myself.

**Comment by gerzon-everneat** [comment]
you are not getting the question, there was no problem at all. there are instances where some features inserted in an endpoint add addtional overhead in the process in which we dont want, say in this case a logger, if we have logger all over the api its kinda slow each request. so having the async logger without putting await kinda removes the additional overhead and still logs the event

**Comment by AhmedBaset** [comment]
https://nextjs.org/docs/app/api-reference/functions/unstable_after
